src/luaclang.c

```diff
@@ -1,14 +1,312 @@
 #include <clang-c/Index.h>
+#include <stdbool.h>
+
 #include "lua.h"
 #include "lualib.h"
 #include "lauxlib.h"
 
-static const struct luaL_Reg luaclang [] = {
-      {NULL, NULL}  /* sentinel */
+#define CLANG_METATABLE  "Clang.Object"
+#define CURSOR_METATABLE "Clang.Cursor"
+#define TYPE_METATABLE   "Clang.Type"
+
+typedef struct clang_object {
+        CXIndex idx;
+        CXTranslationUnit tu;
+} clang_object;
+
+/* Create clang_object as userdata */
+static clang_object *new_clangobject(lua_State *L) 
+{
+        clang_object *clang_parser = (clang_object*) lua_newuserdata(L, sizeof(clang_object));
+        luaL_getmetatable(L, CLANG_METATABLE);
+        lua_setmetatable(L, -2);
+        return clang_parser;
+}
+
+/* Convert userdata type to  */
+static clang_object to_clangobject(lua_State *L, int n) 
+{
+        clang_object *clang_parser = (clang_object*) luaL_checkudata(L, n, CLANG_METATABLE);
+        return *clang_parser;
+}
+
+/*      
+        Format - newParser(file_name)
+        Parameter - file_name - The name of the source file to load 
+        More info - 1. https://clang.llvm.org/doxygen/group__CINDEX.html#ga51eb9b38c18743bf2d824c6230e61f93
+                    2. https://clang.llvm.org/doxygen/group__CINDEX__TRANSLATION__UNIT.html#ga2baf83f8c3299788234c8bce55e4472e
+        Returns clang object whose translation unit cursor can be obtained.
+*/
+static int create_clangobject(lua_State *L)
+{
+        const char *file_name = lua_tostring(L, 1);
+        clang_object *clang_parser = new_clangobject(L);
+        clang_parser->idx = clang_createIndex(1, 0);
+        const char *args[] = {file_name};
+        clang_parser->tu = clang_parseTranslationUnit(clang_parser->idx, 0, args, 1, 0, 0, CXTranslationUnit_None);
+        return 1;
+}
+
+/*      
+        Format - parserdisposeParser()
+        Parameter - parser - Clang object to be disposed
+        More info - 1. https://clang.llvm.org/doxygen/group__CINDEX.html#ga51eb9b38c18743bf2d824c6230e61f93
+                    2. https://clang.llvm.org/doxygen/group__CINDEX__TRANSLATION__UNIT.html#ga2baf83f8c3299788234c8bce55e4472e
+        Returns nothing
+*/
+static int dispose_clangobject(lua_State *L)
+{
+        clang_object clang_parser = to_clangobject(L, 1);
+        if (clang_parser.idx == NULL) return 0;
+        clang_disposeIndex(clang_parser.idx);
+        clang_disposeTranslationUnit(clang_parser.tu);
+        clang_parser.idx = NULL;
```

Comment:
This assignment only modifies the local variable; it doesn't modify the userdata that Lua has stored.